### PR TITLE
Create vacation request

### PR DESCRIPTION
**Pull request by eleidan**
Major parts of the form are implemented + Updates according to comments

TODO:
- Consider day offs and weekends [`postponed`]
- Consider proper handling of available days vs requested days [`postponed`]
- Write some tests [`postponed`]
- Other improvements, to be agreed on

### CODE REVIEW

File: app/assets/javascripts/models/available_vacation.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+App.Models.AvailableVacation = Backbone.Model.extend({
```
Comment (epmlys) [inline app/assets/javascripts/models/available_vacation.js]: Looks like this model does nothing yet. Please remove it.

File: app/assets/javascripts/models/available_vacation.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+  defaults: {
```
Comment (epmlys) [inline app/assets/javascripts/models/available_vacation.js]: Empty defaults should be removed.

File: app/assets/javascripts/routers/router.js [hunk 3]
```diff
@@ -24,6 +25,11 @@ App.Router = Backbone.Router.extend({
     $('.content').html("Just a stub for the future Calendar...");
   },
 
+  vacation_requests: function() {
+    console.log('vacation_requests');
```
Comment (epmlys) [inline app/assets/javascripts/routers/router.js]: Any debug messages should be removed before creating Pull Request.

File: app/assets/javascripts/routers/router.js [hunk 4]
```diff
@@ -24,6 +25,11 @@ App.Router = Backbone.Router.extend({
...
+    $('.content').html("You should see the <strong>Request</strong> form... that's all I can tell you, dude O_O");
+    App.vacation_requests = new App.Views.VacationRequest();
+  }
   // default: function() {
```
Comment (epmlys) [inline app/assets/javascripts/routers/router.js]: The same with commented code, please remove here and through the code bellow.

File: app/assets/javascripts/views/vacation_request.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
+App.Views.VacationRequest = Backbone.View.extend({
+  el: 'section',
+  template: JST['templates/vacation_request'],
+
+  initialize: function() {
+    this.$el.html( this.template() );
```
Comment (epmlys) [inline app/assets/javascripts/views/vacation_request.js]: In Javascript extra space after '(' and before ')' not required, it's Ruby style. Please remove.

File: app/assets/javascripts/views/vacation_request.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    this.updateAvailableDays();
+    this.updateOkButton();
+
+    // TODO: Disable option as described below when there is no days available
+    // http://www.mkyong.com/jquery/how-to-set-a-dropdown-box-value-in-jquery/
+
+    return this;
+  },
+
+  events: {
```
Comment (epmlys) [inline app/assets/javascripts/views/vacation_request.js]: Events mapping usually goes before initialize near the top of view. You can move it there.

File: app/assets/javascripts/views/vacation_request.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  initialize: function() {
+    this.$el.html( this.template() );
+    // Prepare handy access for the controls
+    this.$available_days  = this.$('#available_days');
+    this.$vacation_type   = this.$('select[name=vacation_type]');
+
+    this.available_vacations  = new App.Collections.AvailableVacations();
+    this.vacation_requests    = new App.Collections.VacationRequests();
+    this.listenTo( this.available_vacations,  'reset',  this.render );
+    this.listenTo( this.vacation_requests,    'all',  this.logg );
```
Comment (epmlys) [inline app/assets/javascripts/views/vacation_request.js]: Just one space after ',', please remove extra spaces.

File: app/assets/javascripts/views/vacation_request.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+  onCancel: function( op ) {
+    console.log('onCancel' + op);
+  },
+
+  logg: function( e, p ) {
+    console.log(e);
+  },
+
+  updateOkButton: function() {
```
Comment (epmlys) [inline app/assets/javascripts/views/vacation_request.js]: It rather means 'disableOkButton', not sure why we need to disable button at all. 
What you do here is validator job, you can look at jquery.validator or something like this.

File: app/assets/javascripts/views/vacation_request.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+  events: {
+    'click  .ok':                         'onCreate',
+    'click  .cancel':                     'onCancel',
+    'change select[name=vacation_type]':  'onTypeChange',
+    'change input[name=start]':           'onStartDate',
+    'change input[name=end]':             'onEndDate',
+  },
+
+  calculateDuration: function( a, b ) {
```
Comment (epmlys) [inline app/assets/javascripts/views/vacation_request.js]: Please give  to a, b parameters more meaningful names, for example 'start', 'end'.

File: app/controllers/vacation_requests_controller.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  # POST
+  # /resource
+  def create
+    @vacation_request = VacationRequest.new vacation_requests_params
+    # Convert JS Date/Time in miliseconds to Ruby Date format
+    @vacation_request.start = Time.at(@vacation_request.start / 1000)
+    @vacation_request.user_id = current_user.id
+    if @vacation_request.save
+      respond_with @vacation_request
+    else
```
Comment (epmlys) [inline app/controllers/vacation_requests_controller.rb]: What is the logic of this else, if do the same here?

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -9,15 +9,16 @@
 </head>
```
Comment (epmlys) [inline app/views/layouts/application.html.erb]: Is it just duplicate of Backbone based UI? Or is this admin part?

File: app/assets/javascripts/views/vacation_request.js (hunk 7 above)
Comment (eleidan) [inline app/assets/javascripts/views/vacation_request.js]: Why? Interpreter does not care about that, but I do. I have no coding style guides that I have to follow.

Comment (epmlys) [inline app/assets/javascripts/views/vacation_request.js]: This one is very simple but covers all the basics - http://javascript.crockford.com/code.html
Also you can install JsLInt and use it with your IDE or editor.

File: app/assets/javascripts/views/vacation_request.js (hunk 8 above)
Comment (eleidan) [inline app/assets/javascripts/views/vacation_request.js]: Actually, it means `updateOkButoonState` =)
Thanks for the `jquery.validator`.

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -2,4 +2,14 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+
+protected
+
+  def authenticate_user!
```
Comment (andriiNyzhnik) [inline app/controllers/application_controller.rb]: Protected, public, private methods shouldn't be nested https://github.com/bbatsov/ruby-style-guide#indent-public-private-protected. Use rubocop for ruby code.

File: app/controllers/available_vacations_controller.rb [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  # GET
+  # /resource
+  # All available vacations for the user
+  def index
+    render json: current_user.available_vacations
+  end
+
+  # POST
+  # /resource
+  def create
```
Comment (andriiNyzhnik) [inline app/controllers/available_vacations_controller.rb]: Create, update, destroy is not used. Please remove them. Also from routes.

File: app/controllers/vacation_requests_controller.rb (hunk 10 above)
Comment (andriiNyzhnik) [inline app/controllers/vacation_requests_controller.rb]: I think there should be:
`flash[:success] = 'Vacation request has been successfully created' if @vacation_request.save`
`respond_with @vacation_request`, according to responders gem.
Also remove comments from generated rails controllers.

**Comment by andriiNyzhnik** [comment]
I suggest to use haml instead of erb, but it is your choice.

File: config/routes.rb [hunk 14]
```diff
@@ -5,5 +5,15 @@
...
+  resources :teams,
+            only: [:index, :create, :edit, :update, :destroy],
+            defaults: { format: :json }
+
+  resources :available_vacations,
+            only: [:index, :create, :edit, :update, :destroy],
+            defaults: { format: :json }
+
+  resources :vacation_requests,
+            only: [:index, :create, :edit, :update, :destroy],
```
Comment (andriiNyzhnik) [inline config/routes.rb]: update and destroy is not used

File: config/routes.rb [hunk 15]
```diff
@@ -5,5 +5,15 @@
     sessions: 'users/sessions'
   }
 
-  resources :teams, defaults: { format: :json }
+  resources :teams,
+            only: [:index, :create, :edit, :update, :destroy],
+            defaults: { format: :json }
+
+  resources :available_vacations,
+            only: [:index, :create, :edit, :update, :destroy],
```
Comment (andriiNyzhnik) [inline config/routes.rb]: create, update, destroy is not used

File: db/schema.rb [hunk 16]
```diff
@@ -11,7 +11,7 @@
 #
```
Comment (andriiNyzhnik) [inline db/schema.rb]: You musn't commit `db/schema.rb`

File: db/seeds.rb [hunk 17]
```diff
@@ -5,3 +5,209 @@
 #
 #   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
 #   Mayor.create(name: 'Emanuel', city: cities.first)
+# https://codedecoder.wordpress.com/2013/04/25/rake-db-seed-in-rails/
```
Comment (andriiNyzhnik) [inline db/seeds.rb]: It is not rails style how you create seed data.

File: db/seeds.rb [hunk 18]
```diff
@@ -5,3 +5,209 @@
 #
```
Comment (andriiNyzhnik) [inline db/seeds.rb]: It is not rails style how you create seed data.

File: app/views/layouts/application.html.erb (hunk 11 above)
Comment (eleidan) [inline app/views/layouts/application.html.erb]: Please emphasize what's wrong here. It just works :)
BB does not care about menu now.

File: app/controllers/application_controller.rb (hunk 12 above)
Comment (eleidan) [inline app/controllers/application_controller.rb]: I don't like this rule.

**Pull request merged**